crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1882 @@
+                index: usize,
+                wait_group: WaitGroup,
+            }
+
+            impl IndexedWaitGroup {
+                async fn wait(self) -> Self {
+                    self.wait_group.wait().await;
+                    self
+                }
+            }
+
+            // Ensure proper backpressure by only polling the buffered iterator when a wait group
+            // is free.
+            let mut wait_groups = (0..num_pipelines)
+                .map(|index| {
+                    let wait_group = WaitGroup::default();
+                    {
+                        let _prime_this_wait_group = wait_group.token();
+                    }
+                    IndexedWaitGroup {
+                        index,
+                        wait_group: WaitGroup::default(),
+                    }
+                    .wait()
+                })
+                .collect::<FuturesUnordered<_>>();
+
+            let mut df_stream = row_group_data_fetcher
+                .into_stream()
+                .map(|x| async {
+                    match x {
+                        Ok(handle) => handle.await,
+                        Err(e) => Err(e),
+                    }
+                })
+                .buffered(row_group_prefetch_size)
+                .map(|x| async {
+                    let row_group_decoder = row_group_decoder.clone();
+
+                    match x {
+                        Ok(row_group_data) => {
+                            async_executor::spawn(TaskPriority::Low, async move {
+                                row_group_decoder.row_group_data_to_df(row_group_data).await
+                            })
+                            .await
+                        },
+                        Err(e) => Err(e),
+                    }
+                })
+                .buffered(
+                    // Because we are using an ordered buffer, we may suffer from head-of-line blocking,
+                    // so we add a small amount of buffer.
+                    num_pipelines + 4,
+                );
+
+            let morsel_seq_ref = &mut MorselSeq::default();
+            let mut dfs = vec![].into_iter();
+
+            'main: loop {
+                let Some(mut indexed_wait_group) = wait_groups.next().await else {
+                    break;
+                };
+
+                if dfs.len() == 0 {
+                    let Some(v) = df_stream.next().await else {
+                        break;
+                    };
+
+                    let v = v?;
+                    assert!(!v.is_empty());
+
+                    dfs = v.into_iter();
+                }
+
+                let mut df = dfs.next().unwrap();
+                let morsel_seq = *morsel_seq_ref;
+                *morsel_seq_ref = morsel_seq.successor();
+
+                loop {
+                    use crate::async_primitives::connector::SendError;
+
+                    let port_index = indexed_wait_group.index;
```

Comment:
Can you give this a different name? A `port` refers to an input or output of a node. One `port` can consist of a serial sender/receiver, or a series of parallel senders/receivers, but they all belong to the same port.

Reply:
I've renamed it to `channel_index`